### ISSUE TITLE
some memory issues

### ISSUE THREAD

**Issue by Saltyptisan** [wei-mao-2019/LearnTrajDep#22]
After multiple calls to loss_func.mpjpe_error and  loss_func.euler_error,  GPU memory is fully occupied. This may be due to some variables not being released properly.

**Comment by Logan-007L** [comment]
Hello, may I ask how do you solve this problem?  I found that when executing quickdemo, the GPU memory occupied by each epoch will be superimposed and not released.  What should I do？thank you !

**Comment by Logan-007L** [comment]
> After multiple calls to loss_func.mpjpe_error and loss_func.euler_error, GPU memory is fully occupied. This may be due to some variables not being released properly.

Hello, may I ask how do you solve this problem? I found that when executing quickdemo, the GPU memory occupied by each epoch will be superimposed and not released. What should I do？thank you !

**Comment by Saltyptisan** [comment]
data_utils.py
~~~
def expmap2rotmat_torch(r):
    """
    Converts expmap matrix to rotation
    batch pytorch version ported from the corresponding method above
    :param r: N*3
    :return: N*3*3
    """
    theta = torch.norm(r, 2, 1)
    r0 = torch.div(r, theta.unsqueeze(1).repeat(1, 3) + 0.0000001)
    r1 = torch.zeros_like(r0).repeat(1, 3)
    # print("In e2r 566, gpu allocated:", torch.cuda.memory_allocated())
    r1[:, 1] = -r0[:, 2]
    r1[:, 2] = r0[:, 1]
    r1[:, 5] = -r0[:, 0]
    r1 = r1.view(-1, 3, 3)
    r1 = r1 - r1.transpose(1, 2)
    n = r1.data.shape[0]
    R = Variable(torch.eye(3, 3).repeat(n, 1, 1)).float().cuda().detach()
    R_2 = torch.mul(torch.sin(theta).unsqueeze(1).repeat(1, 9).view(-1, 3, 3), r1).detach()
    R_3 = torch.mul((1 - torch.cos(theta).unsqueeze(1).repeat(1, 9).view(-1, 3, 3)), torch.matmul(r1, r1)).detach()
    R = R + R_2 + R_3

    # del R,r1,r0
    print("expmap2rotmat_torch :",R.size(),r1.size(),r0.size())
    return R
~~~
改了 R 的计算式，但不太明白原因，猜测是R的梯度矩阵被留在内存中了~

**Comment by Saltyptisan** [comment]
> > After multiple calls to loss_func.mpjpe_error and loss_func.euler_error, GPU memory is fully occupied. This may be due to some variables not being released properly.
> 
> Hello, may I ask how do you solve this problem? I found that when executing quickdemo, the GPU memory occupied by each epoch will be superimposed and not released. What should I do？thank you !

data_utils.py
~~~
def expmap2rotmat_torch(r):
    """
    Converts expmap matrix to rotation
    batch pytorch version ported from the corresponding method above
    :param r: N*3
    :return: N*3*3
    """
    theta = torch.norm(r, 2, 1)
    r0 = torch.div(r, theta.unsqueeze(1).repeat(1, 3) + 0.0000001)
    r1 = torch.zeros_like(r0).repeat(1, 3)
    # print("In e2r 566, gpu allocated:", torch.cuda.memory_allocated())
    r1[:, 1] = -r0[:, 2]
    r1[:, 2] = r0[:, 1]
    r1[:, 5] = -r0[:, 0]
    r1 = r1.view(-1, 3, 3)
    r1 = r1 - r1.transpose(1, 2)
    n = r1.data.shape[0]
    
    R = Variable(torch.eye(3, 3).repeat(n, 1, 1)).float().cuda().detach()
    R_2 = torch.mul(torch.sin(theta).unsqueeze(1).repeat(1, 9).view(-1, 3, 3), r1).detach()
    R_3 = torch.mul((1 - torch.cos(theta).unsqueeze(1).repeat(1, 9).view(-1, 3, 3)), torch.matmul(r1, r1)).detach()
    R = R + R_2 + R_3

    return R
~~~
改了R的式子，
我猜测是计算R的每个过程都被保留了梯度矩阵，但实际上又用不到~

**Comment by Logan-007L** [comment]
> > > After multiple calls to loss_func.mpjpe_error and loss_func.euler_error, GPU memory is fully occupied. This may be due to some variables not being released properly.
> > 
> > 
> > Hello, may I ask how do you solve this problem? I found that when executing quickdemo, the GPU memory occupied by each epoch will be superimposed and not released. What should I do？thank you !
> 
> data_utils.py
> 
> ```
> def expmap2rotmat_torch(r):
>     """
>     Converts expmap matrix to rotation
>     batch pytorch version ported from the corresponding method above
>     :param r: N*3
>     :return: N*3*3
>     """
>     theta = torch.norm(r, 2, 1)
>     r0 = torch.div(r, theta.unsqueeze(1).repeat(1, 3) + 0.0000001)
>     r1 = torch.zeros_like(r0).repeat(1, 3)
>     # print("In e2r 566, gpu allocated:", torch.cuda.memory_allocated())
>     r1[:, 1] = -r0[:, 2]
>     r1[:, 2] = r0[:, 1]
>     r1[:, 5] = -r0[:, 0]
>     r1 = r1.view(-1, 3, 3)
>     r1 = r1 - r1.transpose(1, 2)
>     n = r1.data.shape[0]
>     
>     R = Variable(torch.eye(3, 3).repeat(n, 1, 1)).float().cuda().detach()
>     R_2 = torch.mul(torch.sin(theta).unsqueeze(1).repeat(1, 9).view(-1, 3, 3), r1).detach()
>     R_3 = torch.mul((1 - torch.cos(theta).unsqueeze(1).repeat(1, 9).view(-1, 3, 3)), torch.matmul(r1, r1)).detach()
>     R = R + R_2 + R_3
> 
>     return R
> ```
> 
> 改了R的式子， 我猜测是计算R的每个过程都被保留了梯度矩阵，但实际上又用不到~

Thanks a lot! Thank you very much! I'm a beginner. What a pleasure to have such help!

**Comment by Saltyptisan** [comment]
> > > > After multiple calls to loss_func.mpjpe_error and loss_func.euler_error, GPU memory is fully occupied. This may be due to some variables not being released properly.
> > > 
> > > 
> > > Hello, may I ask how do you solve this problem? I found that when executing quickdemo, the GPU memory occupied by each epoch will be superimposed and not released. What should I do？thank you !
> > 
> > 
> > data_utils.py
> > ```
> > def expmap2rotmat_torch(r):
> >     """
> >     Converts expmap matrix to rotation
> >     batch pytorch version ported from the corresponding method above
> >     :param r: N*3
> >     :return: N*3*3
> >     """
> >     theta = torch.norm(r, 2, 1)
> >     r0 = torch.div(r, theta.unsqueeze(1).repeat(1, 3) + 0.0000001)
> >     r1 = torch.zeros_like(r0).repeat(1, 3)
> >     # print("In e2r 566, gpu allocated:", torch.cuda.memory_allocated())
> >     r1[:, 1] = -r0[:, 2]
> >     r1[:, 2] = r0[:, 1]
> >     r1[:, 5] = -r0[:, 0]
> >     r1 = r1.view(-1, 3, 3)
> >     r1 = r1 - r1.transpose(1, 2)
> >     n = r1.data.shape[0]
> >     
> >     R = Variable(torch.eye(3, 3).repeat(n, 1, 1)).float().cuda().detach()
> >     R_2 = torch.mul(torch.sin(theta).unsqueeze(1).repeat(1, 9).view(-1, 3, 3), r1).detach()
> >     R_3 = torch.mul((1 - torch.cos(theta).unsqueeze(1).repeat(1, 9).view(-1, 3, 3)), torch.matmul(r1, r1)).detach()
> >     R = R + R_2 + R_3
> > 
> >     return R
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 改了R的式子， 我猜测是计算R的每个过程都被保留了梯度矩阵，但实际上又用不到~
> 
> Thanks a lot! Thank you very much! I'm a beginner. What a pleasure to have such help!

You're welcome. Happy to help~

**Comment by Logan-007L** [comment]
I am sorry to trouble you again. I successfully ran the training code and found that it takes about 20 hours to train 50 epochs on the h3.6m dataset. Is that the same for you? The training was taking so long that I was wondering if my server was having a problem

**Comment by xiaoxinhu** [comment]
> 很抱歉再次打扰你。我成功运行了训练代码，发现在 h3.6m 数据集上训练了 50 个 epoch 大约需要 20 个小时。你的情况也是这样吗？训练运行太长了，我怀疑不是我的服务器产生的问题

您好 您解决这个问题了吗 我也运行要好久
